src/protocols/grpc/hpr_grpc_client_connection_pool.erl

```diff
@@ -0,0 +1,292 @@
+                ReservedConnectionRecord#reserved_connection.reservation_ref,
+                IdleTimeout
+            )
+    end.
+
+-spec maybe_start_idle_timer(
+    Lsn :: lns(),
+    IdleTimeout :: idle_timeout(),
+    ReferenceCount :: non_neg_integer()
+) -> ok.
+% Start an idle timer if the connection reference count is 0
+maybe_start_idle_timer(Lsn, IdleTimeout, 0) ->
+    set_idle_timer(Lsn, IdleTimeout);
+maybe_start_idle_timer(_Lsn, _, _) ->
+    ok.
+
+% returns new value
+-spec connection_reference_count(lns(), integer()) -> integer().
+connection_reference_count(Lns, Increment) ->
+    ets:update_counter(
+        connection,
+        Lns,
+        {#connection.reference_count, Increment}
+    ).
+
+-spec grpc_client_connect(endpoint()) ->
+    {ok, grpc_client:connection()} | {error, any()}.
+grpc_client_connect({Transport, Host, Port}) ->
+    grpc_client:connect(Transport, Host, Port, []).
+
+-spec insert_connection
+    (lns(), {ok, grpc_client:connection()}) -> {ok, grpc_client:connection()};
+    (lns(), {error, any()}) -> {error, any()}.
+insert_connection(Lns, {ok, GrpcClientConnection}) ->
+    ets:insert(connection, #connection{
+        lns = binary:copy(Lns),
+        connection = GrpcClientConnection
+    }),
+    {ok, GrpcClientConnection};
+insert_connection(_, {error, Error}) ->
+    {error, Error}.
+
+-spec cancel_idle_timer(#connection{}) -> ok.
+cancel_idle_timer(#connection{idle_timeout_timer_ref = undefined}) ->
+    ok;
+cancel_idle_timer(ConnectionRecord) ->
+    timer:cancel(ConnectionRecord#connection.idle_timeout_timer_ref),
+    ets:update_element(connection, ConnectionRecord#connection.lns, {
+        #connection.idle_timeout_timer_ref, undefined
+    }),
+    ok.
+
+-spec insert_reserved_connection(lns(), reservation_ref(), monitor_ref()) -> ok.
+insert_reserved_connection(Lns, ReservationRef, MonitorRef) ->
+    ets:insert(
+        reserved_connection,
+        #reserved_connection{
+            reservation_ref = ReservationRef,
+            lns = binary:copy(Lns),
+            monitor_ref = MonitorRef
+        }
+    ),
+    ok.
+
+-spec set_idle_timer(lns(), idle_timeout()) -> ok.
+set_idle_timer(_Lsn, infinity) ->
+    ok;
+set_idle_timer(Lsn, IdleTimeout) ->
+    TimerRef = timer:send_after(IdleTimeout, {timeout, Lsn}),
+    ets:update_element(
+        connection,
+        Lsn,
+        {#connection.idle_timeout_timer_ref, TimerRef}
+    ),
+    ok.
+
+-spec monitor_owner(pid(), reservation_ref()) -> monitor_ref().
+monitor_owner(Owner, ReservationRef) ->
+    erlang:monitor(process, Owner, [{tag, {'DOWN', ReservationRef}}]).
+
+-spec demonitor_owner(monitor_ref()) -> ok.
+demonitor_owner(MonitorRef) ->
+    erlang:demonitor(MonitorRef),
+    ok.
+
+-spec decode_lns(lns()) -> endpoint().
+decode_lns(Lns) ->
+    {Address, Port} = decode_lns_parts(binary:split(Lns, <<$:>>)),
+    {tcp, Address, Port}.
+
+-spec decode_lns_parts([binary()]) -> {string(), non_neg_integer()} | none().
```

Comment:
There's a version of this in `hpr_gwmp_router`. 
Don't know if it makes sense to reuse it here.

Reply:
I think we're replacing these with library functions in a lorawan utility library, once available.